File: in_Java/0456 132 Pattern.java
class Solution {
    public boolean find132pattern(int[] nums) {
        Stack<Integer> st = new Stack<>();
        int curmin = Integer.MAX_VALUE;
        for(int i : nums){
            if(i <= curmin){
                curmin = i;
            }else{
                while(!st.isEmpty()){
                    if(st.peek() >= i) break;
                    st.pop(); // pop away smaller value
                    if(st.pop() > i) return true; // after popping smaller value, check for Ak
                }
                st.push(i);
                st.push(curmin);
            }
        }
        return false;
    }
}

File: in_Python/0456 132 Pattern.py
class Solution(object):
    def find132pattern(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        if nums == None or len(nums) < 3:
            return False
        minarr = [-1] * len(nums)
        minarr[0] = nums[0]
        for i in range(1, len(nums)):
            minarr[i] = min(minarr[i-1], nums[i])
        st = []
        for j in range(len(nums)-1, -1, -1):
            if nums[j] <= minarr[j]:
                continue
            while st and st[-1] <= minarr[j]:
                st.pop()
            if st and st[-1] < nums[j]:
                return True
            st.append(nums[j])
        return False
            